modules/siddhi-core/src/main/java/io/siddhi/core/table/record/AbstractQueryableRecordTable.java

```diff
@@ -431,103 +335,149 @@ public void updateOrAdd(ComplexEventChunk<StateEvent> updateOrAddingEventChunk,
+                        matchingEvent);
+                return cacheResults;
             } finally {
-                readWriteLock.writeLock().unlock();
-            }
-            if (cachedTable.size() > maxCacheSize) {
-                isCacheEnabled = false;
-                log.warn(siddhiAppContext.getName() + ": " + cacheTableDefinition.getId() + " size is now " +
-                        cachedTable.size() + " which is bigger than cache table size defined as " + maxCacheSize +
-                        ". So cache is now disabled");
+                readWriteLock.readLock().unlock();
             }
         } else {
-            if (recordTableHandler != null) {
-                recordTableHandler.updateOrAdd(timestamp, recordStoreCompiledCondition.compiledCondition,
-                        updateConditionParameterMaps, recordTableCompiledUpdateSet.getUpdateSetMap(),
-                        updateSetParameterMaps, addingRecords);
-            } else {
-                updateOrAdd(recordStoreCompiledCondition.compiledCondition, updateConditionParameterMaps,
-                        recordTableCompiledUpdateSet.getUpdateSetMap(), updateSetParameterMaps, addingRecords);
+            if (log.isDebugEnabled()) {
+                log.debug(siddhiAppContext.getName() + ": store table size is bigger than cache.");
             }
-        }
+            if (cacheEnabled && compiledConditionWithCache.isRouteToCache()) {
+                if (log.isDebugEnabled()) {
+                    log.debug(siddhiAppContext.getName() + ": cache constraints satisfied. Checking cache");
+                }
+                readWriteLock.readLock().lock();
+                try {
+                    cacheResults = cacheTable.find(compiledConditionWithCache.getCacheCompileCondition(),
+                            matchingEvent);
+                    if (cacheResults != null) {
+                        if (log.isDebugEnabled()) {
+                            log.debug(siddhiAppContext.getName() + ": cache hit. Sending results from cache");
+                        }
+                        return cacheResults;
+                    }
+                } finally {
+                    readWriteLock.readLock().unlock();
+                }
+                // cache miss
+                if (log.isDebugEnabled()) {
+                    log.debug(siddhiAppContext.getName() + ": cache miss. Loading from store");
+                }
+                StreamEvent streamEvent = super.find(recordStoreCompiledCondition, matchingEvent);
+                if (streamEvent == null) {
+                    if (log.isDebugEnabled()) {
+                        log.debug(siddhiAppContext.getName() + ": store also miss. sending null");
+                    }
+                    return null;
+                }
 
+                if (cacheTable.size() == maxCacheSize) {
+                    ((CacheTable) cacheTable).deleteOneEntryUsingCachePolicy();
```

Comment:
don't we need to protect cache table with lock what if there are multiple concurrent writes? the read-write lock is not used here